Add basic dash

FerchusGames opened the pull request:
Hice otra interfaz para eventos que se comunican directamente del PlayerManager al PlayerController, como DashEnd.

FerchusGames committed:
Changed Move observable from Vector2 to float

FerchusGames committed:
Added dash movement with start and end events

Horothenic commented on Towerfall/Assets/_Main/Scripts/Player/PlayerManager.cs:
```diff
@@ -7,35 +8,61 @@
+        public IObservable<float> JumpStart => _jumpStartSubject.AsObservable();
+        public IObservable<float> Move => _moveSubject.AsObservable();
+        public IObservable<Vector2> DashStart => _dashStartSubject.AsObservable();
+
```
Remove this line, is OK for them to be together.

FerchusGames replied:
b76a32867a9490e2a7173cbf4ca256d92da7b71f

Horothenic commented on Towerfall/Assets/_Main/Scripts/Player/PlayerManager.cs:
```diff
@@ -7,35 +8,61 @@
+        private void PassMoveEvent(float moveDirection)
         {
-            moveDirection.x *= _playerProperties.MoveAccelerationRate;
+            moveDirection *= _playerProperties.MoveAccelerationRate;
             _moveSubject.OnNext(moveDirection);
```
Move the operation to `_moveSubject.OnNext();` so we keep as less lines as possible, in this case is desired cause is very simple logic, if it gets messier then we can add more variables here.

FerchusGames replied:
212ca7af1e15d2e6e2dce944c83b800268220451

Horothenic commented on Towerfall/Assets/_Main/Scripts/Player/PlayerManager.cs:
```diff
@@ -7,35 +8,61 @@
-            _playerInput.Jump.Subscribe(PassJumpEvent);
+            _playerInput.JumpStart.Subscribe(PassJumpStartEvent);
             _playerInput.Move.Subscribe(PassMoveEvent);
+            _playerInput.DashStart.Subscribe(PassDashStartEvent);
```
Given how this is evolving I think the `IPlayerInput` observables should be named differently, maybe `Action`, so it would be `JumpStartAction`, this will make it less confusing since we are starting to use the same names here.

FerchusGames replied:
d7dd0db43374ee04a601b14ef99a2961c364a854

Horothenic commented on Towerfall/Assets/_Main/Scripts/Player/PlayerManager.cs:
```diff
@@ -7,35 +8,61 @@
+    {
+        private async UniTaskVoid Dash(Vector2 dashDirection, float dashDuration)
+        {
+            dashDirection *= _playerProperties.DashForceMagnitude;
+            _dashStartSubject.OnNext(dashDirection);
```
Move the calculation to `OnNext()`.

FerchusGames replied:
fcb552f7a7fc4e1167e2e07af16c442ffebb12dc

Horothenic commented on Towerfall/Assets/_Main/Scripts/Player/PlayerManager.cs:
```diff
@@ -7,35 +8,61 @@
+        public IObservable<Unit> DashEnd => _dashEndSubject.AsObservable();
+
+
+        private void PassJumpStartEvent()
```
Lets change the word `Pass` to `Execute`, I think it will make more sense.

FerchusGames replied:
8ccbff66aed7619d4ec7407697bb75ef783d6966

Horothenic commented on Towerfall/Assets/_Main/Scripts/Player/PlayerManager.cs:
```diff
@@ -7,35 +8,61 @@
+
+        private void PassDashStartEvent(Vector2 dashDirection)
+        {
+            Dash(dashDirection, _playerProperties.DashDuration).Forget();
```
You can remove the parameter from here and use it directly, it will reduce complexity, we should only do this when the value is going to be transformed or the access route is very long, this is because we can access player properties super quieck in all this script.

FerchusGames replied:
fee137a8609414e113281f151214a9cf028bdef9

FerchusGames committed:
Removed extra line

FerchusGames committed:
Moved moveDirection operation directly to parameter

FerchusGames committed:
Changed IPlayerInput observable names

Added "Action" at the end to make it more clear that they come from the InputManager

FerchusGames committed:
Moved dashDirection operation to OnNext() parameter

FerchusGames committed:
Changed input action bridge method names from "Pass" to "Execute"

FerchusGames committed:
Removed dashDuration float from Dash task parameters

We can get the duration directly from the PlayerProperties
